flocker/node/agents/test/blockdevicefactory.py

```diff
@@ -0,0 +1,195 @@
+            'using the CLOUD_CONFIG_FILE environment variable. '
+            'See: '
+            'https://docs.clusterhq.com/en/latest/gettinginvolved/acceptance-testing.html '  # noqa
+            'for details of the expected format.'
+        )
+
+    with open(config_file_path) as config_file:
+        config = safe_load(config_file.read())
+
+    section = config[provider.name]
+    cls, get_kwargs = _BLOCKDEVICE_TYPES[provider]
+    kwargs = dict(cluster_id=uuid4())
+    kwargs.update(get_kwargs(**section))
+    return cls, kwargs
+
+
+def _rackspace_session(username, key, **kwargs):
+    """
+    Create a Keystone session capable of authenticating with Rackspace.
+
+    :param unicode keyname: A RackSpace API username.
+    :param unicode key: A RackSpace API key.
+
+    :return: A ``keystoneclient.session.Session``.
+    """
+    auth = RackspaceAuth(
+        auth_url=RACKSPACE_AUTH_URL,
+        username=username,
+        api_key=key
+    )
+    return Session(auth=auth)
+
+
+def _openstack(region, **config):
+    """
+    Create Cinder and Nova volume managers suitable for use in the creation of
+    a ``CinderBlockDeviceAPI``.
+
+    :param bytes region: The name of the region to which to connect.
+    :param config: Any additional configuration (possibly provider-specific)
+        necessary to authenticate a session for use with the CinderClient and
+        NovaClient.
+
+    :return: A ``dict`` giving initializer arguments for
+        ``CinderBlockDeviceAPI``.
+    """
+    # TODO: Look up the right session factory in the config and use it here
+    # instead of assuming Rackspace.
+    session = _rackspace_session(**config)
+    cinder_client = CinderClient(
+        session=session, region_name=region, version=1
+    )
+    nova_client = NovaClient(
+        session=session, region_name=region, version=2
+    )
+    return dict(
+        cinder_volume_manager=cinder_client.volumes,
+        nova_volume_manager=nova_client.volumes,
+    )
+
+
+def _aws(**config):
+    """
+    Create an EC2 client suitable for use in the creation of an
+    ``EBSBlockDeviceAPI``.
+
+    See ``flocker.node.agents.ebs.ec2_client`` for parameter documentation.
+    """
+    return dict(
+        ec2_client=ec2_client(**config),
+    )
+
+
+_BLOCKDEVICE_TYPES = {
+    ProviderType.openstack: (CinderBlockDeviceAPI, _openstack),
+    ProviderType.aws: (_aws, EBSBlockDeviceAPI),
+}
+
+# ^^^^^^^^^^^^^^^^^^^^ generally useful implementation code, put it somewhere
+# nice and use it
+#
+# https://clusterhq.atlassian.net/browse/FLOC-1840
+#
+# vvvvvvvvvvvvvvvvvvvv testing helper that actually belongs in this module
+
+
+def get_blockdeviceapi_with_cleanup(test_case, provider):
+    """
+    Instantiate an ``IBlockDeviceAPI`` implementation appropriate to the given
+    provider and configured to work in the current environment.  Arrange for
+    all volumes created by it to be cleaned up at the end of the current test
+    run.
```

Comment:
It will also delete any existing volumes. I guess this *shouldn't* be a problem, since we use a fresh cluster id.